### ISSUE TITLE
Exposing kernel pointers to user mode apps via map keys

### ISSUE THREAD

**Issue by dthaler** [vbpf/ebpf-verifier#175]
Same as issue #170 except that 170 did so via map values whereas this issue is about map keys.
Sample program is in https://github.com/vbpf/ebpf-samples/pull/4

A closely related issue is that a program can use an array index that is out of bounds.
For an array map, the verifier should really validate that the key is a number in the range [0, max_entries-1].

**Comment by elazarg** [comment]
Regarding the array map:

* It is straightforward to solve the underflow case if key size is 64bit, since we only track exact content for 64bit writes.
* Checking for overflow requires passing `max_entries` information, which I assume is not really an issue.
* Tracking 32bit writes may result in some slowdown. Additionally, if I am not mistaken, the MSB part becomes "dirty" and a mask or left-right shift will be required. Checking for that would require some adjustment in the abstract domain - perhaps a flag for each register. I think it's related to #150 and #63 (please correct me if this is not true).
* I assume sizes other than 32bit and 64bit do not make sense in array map.

**Comment by dthaler** [comment]
> * I assume sizes other than 32bit and 64bit do not make sense in array map.

According to https://www.man7.org/linux/man-pages/man2/bpf.2.html array map key size must always be 32bit.

**Comment by dthaler** [comment]
This issue covers three test cases: exposeptr2.o, mapoverflow.o, and mapunderflow.o.
The first one was fixed in PR #178.  The other two are still accepted by the verifier and shouldn't be.

**Comment by dthaler** [comment]
@elazarg @caballa I looked into the array index overflow issue some more and started trying to implement such checks.  But then I ran into existing test cases like:
ebpf-samples\linux\lathist_kern.o kprobe/trace_preempt_off

The code for the above eBPF program is:
```
       0:       r0 = get_smp_processor_id:8()
       1:       *(u32 *)(r10 - 4) = r0
       2:       r2 = r10
       3:       r2 += -4
       4:       r1 = map_fd 2052
       6:       r0 = map_lookup_elem:1(r1:FD, r2:K)
       7:       r6 = r0
       8:       if r6 == 0 goto +2 <11>
       9:       r0 = ktime_get_ns:5()
      10:       *(u64 *)(r6 + 0) = r0
      11:       r0 = 0
      12:       exit
```
The max_entries for the array is 4 in the maps section.  In the code above, the key is obtained from the results of get_smp_processor_id() which according to https://people.netfilter.org/rusty/unreliable-guides/kernel-hacking/routines-processorids.html returns a value between 0 and 32.   So in a sense the code is not safe to run on a machine that can return an id of 4 or more, but the program passes the verifier.

It seems that the map helpers are assumed to internally validate the index as being in range and return an error if not, so that the verifier can ignore having to do range bounds checking.   That said, it might be less bug prone for a developer if the verifier actually did bounds checking so the code above could generate a warning that might lead the developer to use a larger max_entries in the map.

Should we
a) not make the verifier do any such checking, so programs that violate the array index bounds should pass verification as they do now
b) make the verifier enforce such checking, so some existing samples that pass will start failing
c) provide a "strict" option that is not the default, so the verifier will do (a) by default but (b) if the developer opts in

I'm leaning towards (c) at the moment but what do others think?

**Comment by elazarg** [comment]
>  That said, it might be less bug prone for a developer if the verifier actually did bounds checking [...]

This is a crucial point: we do _not_ try to help developers write correct programs. We don't even know what a correct program is. The goal of this system is to maximize the number of useful programs, where usefulness in this case is tied to efficiency. The _only_ relevant reason to not perform bounds check is efficiency.

That is why the idiom for packet access was introduced - since calling a dedicated function for each and every memory access is likely to result in very bad performance.

That is also (obviously) why we avoid instrumenting the code with runtime type information / taint analysis to track the kind and origins of every single value in the program, and instead perform it statically in the verifier. Doing this at runtime would make the runtime incredibly slow and so would render the entire system useless.

I am not much of a systems' developer, but it seems that the overhead for a bounds check on a call to map lookup - which is already nontrivial - is negligible, maybe practically zero; and it looks like the Linux devs agree. Similarly to division by zero, as discussed in #133.

There is an obvious tradeoff in the design of the system between runtime overhead and the verification accuracy, and a careful balance is require. What we should _not_ try to do is to find bugs.

That being said, if and when we do introduce verification checks - and therefore false positives - we should also try to help the developer to pinpoint and work around efficiently. I believe this final point aligns with what Arie has expressed in previous discussions.

To summarize:
1) When runtime check has no significant overhead, it is better; it allows more useful programs to run.
2) When such a runtime check has non-negligible cost, and there is a potential safety issue, the verifier should catch it, with as useful error message as possible, and as few false positives as possible.
3) Program correctness issues (including UBs as in #130) are entirely on the developer.

It should also be possible for the verifier to pass information to the runtime, so the AOT compiler can replace expensive operations with cheaper ones (for example, eliminate the runtime check from `map_lookup_elem` in the common case where it is a known constant).

**Comment by elazarg** [comment]
So IMO option (a) is the right way to go. Option (c) is fine if it doesn't introduce too much complexity.

**Comment by caballa** [comment]
We should do what the Linux verifier does. If it does (a) we should do (a). I agree that having a flag with more strict checks (option c) might be useful but not sure if we want to go in that route.

**Comment by dthaler** [comment]
I tested with --domain=linux and the Linux verifier passes (allows) the mapunderflow.o and mapoverflow.o cases, so it does (a).  However, option (c) (and b) is actually pretty simple if we want to go that route.

**Comment by dthaler** [comment]
https://github.com/dthaler/ebpf-verifier/pull/15/commits/2d3019215f370509078d25f66d4da9fa47fb253d shows the changes for option (c) so you can get an idea of the complexity.  Most changes are just to add the option, the real check is just the change in ebpf_domain.hpp

Let me know if you want to go this route (i.e., c) or if we should instead just close this issue and ignore the change shown.

**Comment by dthaler** [comment]
@elazarg @caballa can you check the link in my previous comment and advise on whether you'd like to see an actual PR opened or just close this issue?

**Comment by elazarg** [comment]
Sorry about the delay. 

I'm -0 on this. If you submit a PR I will merge it, but I don't really see the benefit. Do you have a more general use case in mind, for bounds checking or strict mode?

**Comment by caballa** [comment]
I don't have a strong opinion. But if we start having more strict checks that the current verifier then it should be the case that we can convince ourselves that developers want those checks and therefore, the current verifier should include those sooner than later. If not then we shouldn't add them. Elazar does not seem convinced so I vote for not having them for now.

**Comment by poornagmsft** [comment]
Hi everyone, I think as more and more complicated programs are written in ebpf and put in production I would like the system to help surface errors [even if false positive] to the developers as early as possible. There are two cases: First is we have a bug in the ebpf program that is harmless to the OS [no crash/corruption/security issue] but the application behavior is unpredictable. Depending on the bug, this may surface several months after deploying so that is certainly something developers care. Second case is we have a bug in the ebpf program that we know the execution context [the OS] is supposed to prevent but due to a bug in the OS, this results in a catastrophic failure in production. The latter is actually not too uncommon in production systems with callouts [executing 3p code] where buggy 3P callout behavior is not handled correctly by the OS and it crashes/or corrupts memory. Either way, personally I will want atleast an option with the verifier to help surface this as early in the development cycle as possible.

**Comment by xfoukas** [comment]
Hi all, having some optional strict checks for things that might fail at runtime could be useful for some of the use cases we are interested in supporting in a Linux-based project we are working on. Some of the functions we want to support have soft real-time requirements, meaning that we would generally like to avoid introducing any unnecessary time overhead. We can certainly write helper functions that validate the index internally and return an error if it is out of bounds. However, if failures were introduced unintentionally, it means that we would either have to abort the running eBPF program and return immediately without doing anything useful or we would need to add extra code to somehow resolve the (unexpected) failure. In both cases, this would add overhead that could have been avoided if the problem was picked up by the verifier. In that sense, having such checks give out warnings (or be enabled with an optional flag) could be a useful feature for the subset of our programs that has the most strict time requirements.

**Comment by dthaler** [comment]
Based on feedback from poornagmsft and xfoukas, it seems there is sufficient justification from developers to submit the pull request for this optional feature, so I have now done so.

**Comment by elazarg** [comment]
As I said, I will accept the feature, but just to reiterate: this is not about catching bugs. This is about enabling programs that would be impossible (or less useful) otherwise, by allowing the trusted runtime to execute code efficiently. This is related to the last point by @xfoukas (as I understand it).
